Squashed version of blueyed pull requests

artm opened the pull request:
This is basically [request #6](https://github.com/Townk/vim-autoclose/pull/6) but rearranged and squashed to make it easier to review. It hides some interim design decisions that were improved upon later and keeps each topic in close to 1 commit.

blueyed committed:
Doc review.

blueyed committed:
Terminal keys handling fixes

(1) Fix vt100/xterm keys aborting insert mode.

This fixes issue #4, by adding an ambiguous mapping so Vim fully
processes the escape sequences of special keys like cursor keys, pos1,
end etc.

The fix is taken from another autoclose plugin by Karl Guertin, which is
available at http://www.vim.org/scripts/script.php?script_id=1849.

(2) Refactor var and map setup (loops), fix Home/End.

Define s:movementKeys and also add Home/End to them.

This makes the code much cleaner and easier to extend, and adding
Home/End fixes using those after triggering pair insertion.

(3) Improve the fix for xterm movement key behavior.

Add s:needspecialkeyhandling boolean and an extra map for xterm key
codes (s:movementKeysXterm), which need the same mapping. They get
mapped to the "proper" names.

I am not sure if this should not just go to somebodys vimrc, but without
autoclose the keys work properly.

- Also add PageUp and PageDown to the list of movement keys.
- Fix mapping for Xterm special keys.
- Do not use `nore` and add a missing space.
- Fix xterm mapping for PageUp/PageDown.
- Special key handling also for linux* terms.
- Handle rxvt/urxvt via needspecialkeyhandling, too.

blueyed committed:
Minor: fix indent/whitespace.

blueyed committed:
Add augroup guard.

This makes resourcing the file easier.

blueyed committed:
Refactoring: dict defaults and popup menu fixes

This uses the method introduces for movementKeys before for all the
settings.

Also improve default pairs: add '<>', '``' and '«»'.

While the latter is not on regular keyboards, it is provided at least on
the Neo keyboard layout.

Also popup menu related fixes:

Set AutoClosePumvisible* to the key itself by default.

This fixes using the movement keys in a menu, e.g. with the following
text:
  foo
  fooo
  fo_

Typing "fo" opens the autocomplpop popup; using the cursor keys there
made the menu disappear and the movement carried out instead.

pumvisible() is handled depending on
the existence of a Pumvisible* setting.

Only Left, Right, PageUp and PageDown are handled as-is in the
completion menu. All other movement keys will leave the menu and need to
flush the buffer.

blueyed committed:
Improve on when to add closing counterparts.

Previously the closing part was just not inserted before any word chars
(`\w`); now it is added before whitespace and any char listed in
AutoClosePairs.

This fixes putting a "(" in front of e.g. "$", so that you do not get
"()$", but "($" instead.

Check for defined pair and check early.

It could happen that a:char was not defined in b:AutoClosePairs (because
it got removed according to the filetype probably).

This commit now checks for the pair to be set.

It also moves the b:AutoCloseOn and s:IsForbidden checks up to the
beginning of the function.

blueyed committed:
Make AutoClosePairs filetype-aware.

Add `reset` argument to DefineVariables which gets enabled on FileType
events and will reset all settings in the buffer namespace.
An additional (experimental) check looks at the filetype and will add
the '||' pair for &ft=ruby.

See
https://github.com/thenoseman/vim-autoclose/commit/128c289d8ab6b52b0ac723dfd25fcb9387fabd85#commitcomment-324879
for the idea.

Remove '<'/'>' map for typoscript and shell.

This is still experimental and needs some refactoring regarding the lookup.

Commit:
Minor: turn off debug mode

This lets g:loaded_AutoClose work again

pjg commented:
This really should be merged. I've used the version from master and was really surprised to find that both ' and " were not autoclosed, and I lost arrows both in insert mode and in fuzzyfinder_textmate. Using @blueyed version solved my problems :)

Commit:
debug via global

this way debug mode can be switched on/off externally and we don't
commit / release it inadvertently.

Commit:
Don't delete random pairs of characters

If hitting backspace between two identical characters that aren't pairs
don't do smart deletion (or rather be smart and only delete one)

artm commented:
Oops, the last two commits weren't supposed to come here, I'll rework this.

artm commented:
this becomes #14, sorry folks
